Document the useFetchClient hook (#15295)
strapi/strapi: docs/docs/core/hooks/use-fetch-client.mdx

---
title: useFetchClient
slug: /hooks/use-fetch-client
description: API reference for the useFetchClient hook in Strapi
tags:
  - hooks
  - fetch-client
---

An abstraction around the axios instance wrapped in a hook. It provides a simple interface to handle API calls to the Strapi backend.

> It handles request cancellation under the hook with `AbortController`, and when the component gets unmounted all the reqeusts that it has made are aborted.

Reviewer: Can you link to MDN `AbortController` please?

> The hook exposes four methods:

Reviewer: This should be it's own heading of `Methods`, with type information imo

- **del(url, config)**: which is a method that given a relative url, will make a DELETE request to the Strapi backend with the configuration passed as second parameter, if needed.

## Implementation details

The fetch instance is running axios under the hook with two request interceptors.

### Base URL

The default URL will be the one defined in the environment variable: `STRAPI_ADMIN_BACKEND_URL`.

### Interceptors

#### Request

The request interceptor adds the following parameters to the header:

```js
{
    Authorization: `Bearer <AUTH_TOKEN>`,
    Accept: 'application/json',
    'Content-Type': 'application/json',
}
```

#### Response

If everything has gone correctly, the response is returned as it comes from the backend. However, if it cointains a **status code of 401**, the authentication details will be removed from
the application storage and the window will be reload

:::caution
Have this in mind if using this hook in pages where the auth token is not available, such as login, because it will create an infinite loop if not treating the error correctly. See section [Troubleshooting](#troubleshooting)
:::

> ### Basic usage

Reviewer: This should be at the top just called `Usage`

---
title: useFetchClient
slug: /hooks/use-fetch-client
description: API reference for the useFetchClient hook in Strapi
tags:

>   - hooks

Reviewer: ```suggestion
  - hooks
  - axios
  - data
```

  - fetch-client
---

## Usage

> The following examples shows a basic way to use the `useFetchClient` hook to make a get request to a strapi backend endpoint:

Reviewer: ```suggestion
The following example shows a basic way to use the `useFetchClient` hook to make a get request to a Strapi backend endpoint:
```

```jsx
import {useState} from "react"
import useFetchClient from '@strapi/admin/admin/src/hooks/useFetchClient';

> const Component () => {

Reviewer 2: ```suggestion
const Component = () => {
```

    const [items, setItems] = useState([]);
    const { get } = useFetchClient();
    const requestURL = "/some-endpoint";

>     const handleGetData = () => {

Reviewer: ```suggestion
    const handleGetData = async () => {
```

        const { data } = await get(requestURL);
        setItems(data.items);
    }

    return(
        <div>
            <div>
                {

>                     items && items.map(item => <h1>{item.name}</h1>)

Reviewer: ```suggestion
                    items && items.map(item => <h2 key={item.uuid}>{item.name}</h2>)
```

You wouldn't have more than one `h1` in a document.

                }
            </div>
            <button onClick={handleGetData}>Show Items</button>
        </div>
    )
}
```

## Methods

> Essentially, this is an abstraction around the axios instance wrapped in a hook. It provides a simple interface to handle API calls to the Strapi backend.

Reviewer: ```suggestion
Essentially, this is an abstraction around the axios instance exposed by a hook. It provides a simple interface to handle API calls to the Strapi backend.
```

> It handles request cancellation under the hook with [`AbortController`](https://developer.mozilla.org/en-US/docs/Web/API/AbortController), and when the component gets unmounted all the reqeusts that it has made are aborted.

Reviewer: ```suggestion
It handles request cancellations inside the hook with an [`AbortController`](https://developer.mozilla.org/en-US/docs/Web/API/AbortController). This is typically triggered when the component is unmounted so all the requests that it is currently making are aborted.
```

The hook exposes four methods:

> - **get(url, config)**: which is a method that given a relative url (string), will make a GET request to the Strapi backend with the configuration (object) passed as second parameter, if needed.

Reviewer: ```suggestion
- **get(url, config)**: requires a relative url (`string`) and makes a `GET` request to the Strapi backend with an optional configuration `object` passed as second parameter.
```

> - **post(url, data, config)**: which is a method that given a relative url (string), will make a POST request with the data (object) provided to the Strapi backend with the configuration passed as third parameter (object), if needed.

Reviewer: ```suggestion
- **post(url, data, config)**: requires a relative url (`string`) and makes a `POST` request with the required data `object` to the Strapi backend with the optional configuration `object` passed as third parameter.
```

> - **put(url, data, config)**: which is a method that given a relative url (string), will make a PUT request with the data (object) provided to the Strapi backend with the configuration passed as third parameter (object), if needed.

Reviewer: ```suggestion
- **put(url, data, config)**: requires a relative url (`string`) and makes a `PUT` request with the required data `object` to the Strapi backend with the optional configuration `object` passed as third parameter.
```

> - **del(url, config)**: which is a method that given a relative url (string), will make a DELETE request to the Strapi backend with the configuration passed as second parameter (object), if needed.

Reviewer: ```suggestion
- **del(url, config)**: requires a relative url (`string`) and makes a `DELETE` request to the Strapi backend with an optional configuration `object` passed as second parameter.
```

## Implementation details

> The fetch instance is running axios under the hook with two request interceptors.

Reviewer: ```suggestion
The following information is the internal additions we've added to the axios instance via two request interceptors. As well as an explanation of the `baseUrl`.
```

### Base URL

The default URL will be the one defined in the environment variable: `STRAPI_ADMIN_BACKEND_URL`.

### Interceptors

#### Request

The request interceptor adds the following parameters to the header:

```js
{
    Authorization: `Bearer <AUTH_TOKEN>`,
    Accept: 'application/json',
    'Content-Type': 'application/json',
}
```

#### Response

> If everything has gone correctly, the response is returned as it comes from the backend. However, if it cointains a **status code of 401**, the authentication details will be removed from

Reviewer: ```suggestion
If everything has works correctly, the response is returned as it comes from the backend. However, if it contains a **status code of 401** the authentication details will be removed from
```

Author: I guess here we should remove "has" before "works"

> the application storage and the window will be reload

Reviewer: ```suggestion
the application storage and the window will be reloaded.
```

:::caution

> Have this in mind if using this hook in pages where the auth token is not available, such as login, because it will create an infinite loop if not treating the error correctly. See section [Troubleshooting](#troubleshooting)

Reviewer: ```suggestion
Have this in mind if using this hook in pages where the auth token is not available, such as `login`, because it will create an infinite loop. See the [Troubleshooting](#troubleshooting) section for more information.
```

:::

## Further Reading

- [axios instance API](https://axios-http.com/docs/instance)
- [AbortController](https://axios-http.com/docs/cancellation)

## Troubleshooting

### Authentication problems

> Trying to access a protected route from a context where the auth token is not available may lead to an infinite loop, because of the response interceptor that

Reviewer: ```suggestion
Trying to access a protected route from a context where the auth token is not available may lead to an infinite loop due to the response interceptor that
```

> reloads the page when obtaining a 401 response. One option available to avoid this from happening is to not considering a 401 response as an error, as follows:

Reviewer: ```suggestion
reloads the page when obtaining a 401 response. One option to avoid this from happening is to not consider a 401 response as an error, see below:
```
